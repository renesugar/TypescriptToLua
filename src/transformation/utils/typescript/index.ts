import * as ts from "typescript";
import { TransformationContext } from "../../context";

export * from "./nodes";
export * from "./types";

// TODO: Move to separate files?

export function hasExportEquals(sourceFile: ts.SourceFile): boolean {
    return sourceFile.statements.some(node => ts.isExportAssignment(node) && node.isExportEquals);
}

/**
 * Search up until finding a node satisfying the callback
 */
export function findFirstNodeAbove<T extends ts.Node>(node: ts.Node, callback: (n: ts.Node) => n is T): T | undefined {
    let current = node;
    while (current.parent) {
        if (callback(current.parent)) {
            return current.parent;
        } else {
            current = current.parent;
        }
    }
}

export function getFirstDeclarationInFile(symbol: ts.Symbol, sourceFile: ts.SourceFile): ts.Declaration | undefined {
    const originalSourceFile = ts.getParseTreeNode(sourceFile) ?? sourceFile;
    const declarations = (symbol.getDeclarations() ?? []).filter(d => d.getSourceFile() === originalSourceFile);

    return declarations.length > 0 ? declarations.reduce((p, c) => (p.pos < c.pos ? p : c)) : undefined;
}

function isStandardLibraryDeclaration(context: TransformationContext, declaration: ts.Declaration): boolean {
    const sourceFile = declaration.getSourceFile();
    if (!sourceFile) {
        return false;
    }

    return context.program.isSourceFileDefaultLibrary(sourceFile);
}

export function isStandardLibraryType(
    context: TransformationContext,
    type: ts.Type,
    name: string | undefined
): boolean {
    const symbol = type.getSymbol();
    if (!symbol || (name ? symbol.name !== name : symbol.name === "__type")) {
        return false;
    }

    // Assume to be lib function if no valueDeclaration exists
    const declaration = symbol.valueDeclaration;
    if (!declaration) {
        return true;
    }

    return isStandardLibraryDeclaration(context, declaration);
}

export function hasStandardLibrarySignature(
    context: TransformationContext,
    callExpression: ts.CallExpression
): boolean {
    const signature = context.checker.getResolvedSignature(callExpression);
    return signature?.declaration ? isStandardLibraryDeclaration(context, signature.declaration) : false;
}

export function inferAssignedType(context: TransformationContext, expression: ts.Expression): ts.Type {
    return context.checker.getContextualType(expression) ?? context.checker.getTypeAtLocation(expression);
}

export function getAllCallSignatures(type: ts.Type): readonly ts.Signature[] {
    return type.isUnion() ? type.types.flatMap(getAllCallSignatures) : type.getCallSignatures();
}

// Returns true for expressions that may have effects when evaluated
export function isExpressionWithEvaluationEffect(node: ts.Expression): boolean {
    return !(ts.isLiteralExpression(node) || ts.isIdentifier(node) || node.kind === ts.SyntaxKind.ThisKeyword);
}
