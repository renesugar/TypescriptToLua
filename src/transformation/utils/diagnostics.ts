import * as ts from "typescript";
import { LuaTarget } from "../../CompilerOptions";
import { createSerialDiagnosticFactory } from "../../utils";
import { AnnotationKind } from "./annotations";

const createDiagnosticFactory = <TArgs extends any[]>(message: string | ((...args: TArgs) => string)) =>
    createSerialDiagnosticFactory((node: ts.Node, ...args: TArgs) => ({
        file: node.getSourceFile(),
        start: node.getStart(),
        length: node.getWidth(),
        messageText: typeof message === "string" ? message : message(...args),
    }));

export const unsupportedNodeKind = createDiagnosticFactory(
    (kind: ts.SyntaxKind) => `Unsupported node kind ${ts.SyntaxKind[kind]}`
);

export const forbiddenForIn = createDiagnosticFactory("Iterating over arrays with 'for ... in' is not allowed.");

export const unsupportedNoSelfFunctionConversion = createDiagnosticFactory((name?: string) => {
    const nameReference = name ? ` '${name}'` : "";
    return (
        `Unable to convert function with a 'this' parameter to function${nameReference} with no 'this'. ` +
        "To fix, wrap in an arrow function, or declare with 'this: void'."
    );
});

export const unsupportedSelfFunctionConversion = createDiagnosticFactory((name?: string) => {
    const nameReference = name ? ` '${name}'` : "";
    return (
        `Unable to convert function with no 'this' parameter to function${nameReference} with 'this'. ` +
        "To fix, wrap in an arrow function, or declare with 'this: any'."
    );
});

export const unsupportedOverloadAssignment = createDiagnosticFactory((name?: string) => {
    const nameReference = name ? ` to '${name}'` : "";
    return (
        `Unsupported assignment of function with different overloaded types for 'this'${nameReference}. ` +
        "Overloads should all have the same type for 'this'."
    );
});

export const decoratorInvalidContext = createDiagnosticFactory("Decorator function cannot have 'this: void'.");

export const annotationInvalidArgumentCount = createDiagnosticFactory(
    (kind: AnnotationKind, got: number, expected: number) => `'@${kind}' expects ${expected} arguments, but got ${got}.`
);

export const extensionCannotConstruct = createDiagnosticFactory(
    "Cannot construct classes with '@extension' or '@metaExtension' annotation."
);

export const extensionCannotExtend = createDiagnosticFactory(
    "Cannot extend classes with '@extension' or '@metaExtension' annotation."
);

export const extensionCannotExport = createDiagnosticFactory(
    "Cannot export classes with '@extension' or '@metaExtension' annotation."
);

export const extensionInvalidInstanceOf = createDiagnosticFactory(
    "Cannot use instanceof on classes with '@extension' or '@metaExtension' annotation."
);

export const extensionAndMetaExtensionConflict = createDiagnosticFactory(
    "Cannot use both '@extension' and '@metaExtension' annotations on the same class."
);

export const metaExtensionMissingExtends = createDiagnosticFactory(
    "'@metaExtension' annotation requires the extension of the metatable class."
);

export const invalidForRangeCall = createDiagnosticFactory((message: string) => `Invalid @forRange call: ${message}.`);

export const luaTableMustBeAmbient = createDiagnosticFactory(
    "Classes with the '@luaTable' annotation must be ambient."
);

export const luaTableCannotBeExtended = createDiagnosticFactory(
    "Cannot extend classes with the '@luaTable' annotation."
);

export const luaTableInvalidInstanceOf = createDiagnosticFactory(
    "The instanceof operator cannot be used with a '@luaTable' class."
);

export const luaTableCannotBeAccessedDynamically = createDiagnosticFactory("@luaTable cannot be accessed dynamically.");

export const luaTableForbiddenUsage = createDiagnosticFactory(
    (description: string) => `Invalid @luaTable usage: ${description}.`
);

export const luaIteratorForbiddenUsage = createDiagnosticFactory(
    "Unsupported use of lua iterator with '@tupleReturn' annotation in for...of statement. " +
        "You must use a destructuring statement to catch results from a lua iterator with " +
        "the '@tupleReturn' annotation."
);

export const unsupportedAccessorInObjectLiteral = createDiagnosticFactory(
    "Accessors in object literal are not supported."
);

export const unsupportedRightShiftOperator = createDiagnosticFactory(
    "Right shift operator is not supported for target Lua 5.3. Use `>>>` instead."
);

const getLuaTargetName = (version: LuaTarget) => (version === LuaTarget.LuaJIT ? "LuaJIT" : `Lua ${version}`);
export const unsupportedForTarget = createDiagnosticFactory(
    (functionality: string, version: Exclude<LuaTarget, LuaTarget.Universal>) =>
        `${functionality} is/are not supported for target ${getLuaTargetName(version)}.`
);

export const unsupportedProperty = createDiagnosticFactory(
    (parentName: string, property: string) => `${parentName}.${property} is unsupported.`
);

export const invalidAmbientIdentifierName = createDiagnosticFactory(
    (text: string) => `Invalid ambient identifier name '${text}'. Ambient identifiers must be valid lua identifiers.`
);

export const unresolvableRequirePath = createDiagnosticFactory(
    (path: string) => `Cannot create require path. Module '${path}' does not exist within --rootDir.`
);

export const unsupportedVarDeclaration = createDiagnosticFactory(
    "`var` declarations are not supported. Use `let` or `const` instead."
);
