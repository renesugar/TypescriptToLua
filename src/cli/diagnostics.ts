import * as ts from "typescript";
import { createSerialDiagnosticFactory, createDiagnosticFactoryWithCode } from "../utils";

export const tstlOptionsAreMovingToTheTstlObject = createSerialDiagnosticFactory((tstl: Record<string, any>) => ({
    category: ts.DiagnosticCategory.Warning,
    messageText:
        'TSTL options are moving to the "tstl" object. Adjust your tsconfig to look like\n' +
        `"tstl": ${JSON.stringify(tstl, undefined, 4)}`,
}));

export const watchErrorSummary = (errorCount: number): ts.Diagnostic => ({
    file: undefined,
    start: undefined,
    length: undefined,
    category: ts.DiagnosticCategory.Message,
    code: errorCount === 1 ? 6193 : 6194,
    messageText:
        errorCount === 1
            ? "Found 1 error. Watching for file changes."
            : `Found ${errorCount} errors. Watching for file changes.`,
});

const createCommandLineError = <TArgs extends any[]>(code: number, getMessage: (...args: TArgs) => string) =>
    createDiagnosticFactoryWithCode(code, (...args: TArgs) => ({ messageText: getMessage(...args) }));

export const unknownCompilerOption = createCommandLineError(
    5023,
    (name: string) => `Unknown compiler option '${name}'.`
);

export const compilerOptionRequiresAValueOfType = createCommandLineError(
    5024,
    (name: string, type: string) => `Compiler option '${name}' requires a value of type ${type}.`
);

export const optionProjectCannotBeMixedWithSourceFilesOnACommandLine = createCommandLineError(
    5042,
    () => "Option 'project' cannot be mixed with source files on a command line."
);

export const cannotFindATsconfigJsonAtTheSpecifiedDirectory = createCommandLineError(
    5057,
    (dir: string) => `Cannot find a tsconfig.json file at the specified directory: '${dir}'.`
);

export const theSpecifiedPathDoesNotExist = createCommandLineError(
    5058,
    (dir: string) => `The specified path does not exist: '${dir}'.`
);

export const compilerOptionExpectsAnArgument = createCommandLineError(
    6044,
    (name: string) => `Compiler option '${name}' expects an argument.`
);

export const argumentForOptionMustBe = createCommandLineError(
    6046,
    (name: string, values: string) => `Argument for '${name}' option must be: ${values}.`
);

export const optionCanOnlyBeSpecifiedInTsconfigJsonFile = createCommandLineError(
    6064,
    (name: string) => `Option '${name}' can only be specified in 'tsconfig.json' file.`
);

export const optionBuildMustBeFirstCommandLineArgument = createCommandLineError(
    6369,
    () => "Option '--build' must be the first command line argument."
);
