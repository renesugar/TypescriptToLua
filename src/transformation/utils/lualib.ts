import * as ts from "typescript";
import * as lua from "../../LuaAST";
import { LuaLibFeature } from "../../LuaLib";
import { getOrUpdate } from "../../utils";
import { TransformationContext } from "../context";

export { LuaLibFeature };

const luaLibFeatures = new WeakMap<TransformationContext, Set<LuaLibFeature>>();
export function getUsedLuaLibFeatures(context: TransformationContext): Set<LuaLibFeature> {
    return getOrUpdate(luaLibFeatures, context, () => new Set());
}

export function importLuaLibFeature(context: TransformationContext, feature: LuaLibFeature): void {
    getUsedLuaLibFeatures(context).add(feature);
}

export function transformLuaLibFunction(
    context: TransformationContext,
    feature: LuaLibFeature,
    tsParent?: ts.Node,
    ...params: lua.Expression[]
): lua.CallExpression {
    importLuaLibFeature(context, feature);
    const functionIdentifier = lua.createIdentifier(`__TS__${feature}`);
    return lua.createCallExpression(functionIdentifier, params, tsParent);
}
