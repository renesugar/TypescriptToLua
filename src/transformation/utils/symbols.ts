import * as ts from "typescript";
import * as lua from "../../LuaAST";
import { getOrUpdate } from "../../utils";
import { TransformationContext } from "../context";
import { markSymbolAsReferencedInCurrentScopes } from "./scope";

const symbolIdCounters = new WeakMap<TransformationContext, number>();
function nextSymbolId(context: TransformationContext): lua.SymbolId {
    const symbolId = (symbolIdCounters.get(context) ?? 0) + 1;
    symbolIdCounters.set(context, symbolId);
    return symbolId as lua.SymbolId;
}

export interface SymbolInfo {
    symbol: ts.Symbol;
    firstSeenAtPos: number;
}

const symbolInfoMap = new WeakMap<TransformationContext, Map<lua.SymbolId, SymbolInfo>>();
const symbolIdMaps = new WeakMap<TransformationContext, Map<ts.Symbol, lua.SymbolId>>();

export function getSymbolInfo(context: TransformationContext, symbolId: lua.SymbolId): SymbolInfo | undefined {
    return getOrUpdate(symbolInfoMap, context, () => new Map()).get(symbolId);
}

export function getSymbolIdOfSymbol(context: TransformationContext, symbol: ts.Symbol): lua.SymbolId | undefined {
    return getOrUpdate(symbolIdMaps, context, () => new Map()).get(symbol);
}

export function trackSymbolReference(
    context: TransformationContext,
    symbol: ts.Symbol,
    identifier: ts.Identifier
): lua.SymbolId | undefined {
    const symbolIds = getOrUpdate(symbolIdMaps, context, () => new Map());

    // Track first time symbols are seen
    let symbolId = symbolIds.get(symbol);
    if (symbolId === undefined) {
        symbolId = nextSymbolId(context);

        symbolIds.set(symbol, symbolId);
        const symbolInfo = getOrUpdate(symbolInfoMap, context, () => new Map());
        symbolInfo.set(symbolId, { symbol, firstSeenAtPos: identifier.pos });
    }

    markSymbolAsReferencedInCurrentScopes(context, symbolId, identifier);

    return symbolId;
}

export function getIdentifierSymbolId(
    context: TransformationContext,
    identifier: ts.Identifier
): lua.SymbolId | undefined {
    const symbol = context.checker.getSymbolAtLocation(identifier);
    if (symbol) {
        return trackSymbolReference(context, symbol, identifier);
    }
}
