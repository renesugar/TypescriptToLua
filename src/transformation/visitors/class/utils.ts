import * as ts from "typescript";
import { TransformationContext } from "../../context";
import { AnnotationKind, getTypeAnnotations } from "../../utils/annotations";

export function isStaticNode(node: ts.Node): boolean {
    return (node.modifiers ?? []).some(m => m.kind === ts.SyntaxKind.StaticKeyword);
}

export function getExtendsClause(node: ts.ClassLikeDeclarationBase): ts.HeritageClause | undefined {
    return (node.heritageClauses ?? []).find(clause => clause.token === ts.SyntaxKind.ExtendsKeyword);
}

export function getExtendedNode(
    context: TransformationContext,
    node: ts.ClassLikeDeclarationBase
): ts.ExpressionWithTypeArguments | undefined {
    const extendsClause = getExtendsClause(node);
    if (!extendsClause) return;

    const superType = context.checker.getTypeAtLocation(extendsClause.types[0]);
    const annotations = getTypeAnnotations(superType);
    if (!annotations.has(AnnotationKind.PureAbstract)) {
        return extendsClause.types[0];
    }
}

export function getExtendedType(
    context: TransformationContext,
    node: ts.ClassLikeDeclarationBase
): ts.Type | undefined {
    const extendedNode = getExtendedNode(context, node);
    return extendedNode && context.checker.getTypeAtLocation(extendedNode);
}
