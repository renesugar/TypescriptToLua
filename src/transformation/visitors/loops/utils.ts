import * as ts from "typescript";
import * as lua from "../../../LuaAST";
import { TransformationContext } from "../../context";
import { performHoisting, popScope, pushScope, ScopeType } from "../../utils/scope";
import { isAssignmentPattern } from "../../utils/typescript";
import { transformAssignment } from "../binary-expression/assignments";
import { transformAssignmentPattern } from "../binary-expression/destructuring-assignments";
import { transformBlockOrStatement } from "../block";
import { transformIdentifier } from "../identifier";
import { checkVariableDeclarationList, transformBindingPattern } from "../variable-declaration";

export function transformLoopBody(
    context: TransformationContext,
    loop: ts.WhileStatement | ts.DoStatement | ts.ForStatement | ts.ForOfStatement | ts.ForInOrOfStatement
): lua.Statement[] {
    pushScope(context, ScopeType.Loop);
    const body = performHoisting(context, transformBlockOrStatement(context, loop.statement));
    const scope = popScope(context);
    const scopeId = scope.id;

    if (!scope.loopContinued) {
        return body;
    }

    const baseResult: lua.Statement[] = [lua.createDoStatement(body)];
    const continueLabel = lua.createLabelStatement(`__continue${scopeId}`);
    baseResult.push(continueLabel);

    return baseResult;
}

export function getVariableDeclarationBinding(
    context: TransformationContext,
    node: ts.VariableDeclarationList
): ts.BindingName {
    checkVariableDeclarationList(context, node);

    if (node.declarations.length === 0) {
        return ts.createIdentifier("____");
    }

    return node.declarations[0].name;
}

export function transformForInitializer(
    context: TransformationContext,
    initializer: ts.ForInitializer,
    block: lua.Block
): lua.Identifier {
    const valueVariable = lua.createIdentifier("____value");

    if (ts.isVariableDeclarationList(initializer)) {
        // Declaration of new variable

        const binding = getVariableDeclarationBinding(context, initializer);
        if (ts.isArrayBindingPattern(binding) || ts.isObjectBindingPattern(binding)) {
            block.statements.unshift(...transformBindingPattern(context, binding, valueVariable));
        } else {
            // Single variable declared in for loop
            return transformIdentifier(context, binding);
        }
    } else {
        // Assignment to existing variable(s)

        block.statements.unshift(
            ...(isAssignmentPattern(initializer)
                ? transformAssignmentPattern(context, initializer, valueVariable)
                : transformAssignment(context, initializer, valueVariable))
        );
    }

    return valueVariable;
}
