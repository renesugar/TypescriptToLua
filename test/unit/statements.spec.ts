import * as util from "../util";
import { unsupportedNodeKind } from "../../src/transformation/utils/diagnostics";

test("Block statement", () => {
    util.testFunction`
        let a = 4;
        { let a = 42; }
        return a;
    `.expectToMatchJsResult();
});

// https://github.com/TypeScriptToLua/TypeScriptToLua/issues/860
test("Unsupported node adds diagnostic", () => {
    util.testModule`
        const a = () => {
            foo: "bar"
        };
    `.expectDiagnosticsToMatchSnapshot([unsupportedNodeKind.code]);
});
