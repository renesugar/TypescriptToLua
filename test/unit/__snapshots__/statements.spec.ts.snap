// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unsupported node adds diagnostic: code 1`] = `
"a = function()
end"
`;

exports[`Unsupported node adds diagnostic: diagnostics 1`] = `"main.ts(3,13): error TSTL: Unsupported node kind LabeledStatement"`;
